src/psyclone/dynamo0p3.py

```diff
@@ -1519,28 +1527,52 @@ def _declare_maps_invoke(self, parent):
-                               kind=api_config.default_kind["integer"],
-                               pointer=True,
-                               entity_decls=[self.dofmap_name(symtab, arg) +
-                                             "(:,:,:) => null()"]))
-            parent.add(DeclGen(parent, datatype="integer",
-                               kind=api_config.default_kind["integer"],
-                               entity_decls=[self.dofmap_size_name(symtab,
-                                                                   arg)]))
-
             stencil_type = arg.descriptor.stencil['type']
+            if stencil_type == "cross2d":
+                parent.add(UseGen(parent, name="stencil_2D_dofmap_mod", only=True,
+                                  funcnames=["stencil_2D_dofmap_type"]))
+                parent.add(TypeDeclGen(parent, pointer=True,
+                                       datatype="stencil_2D_dofmap_type",
+                                       entity_decls=[map_name + " => null()"]))
+                parent.add(DeclGen(parent, datatype="integer",
+                                   kind=api_config.default_kind["integer"],
+                                   pointer=True,
+                                   entity_decls=[self.dofmap_name(symtab, arg) +
+                                                 "(:,:,:,:) => null()"]))
+                parent.add(DeclGen(parent, datatype="integer",
+                                   kind=api_config.default_kind["integer"],
+                                   pointer=True,
+                                   entity_decls=[self.dofmap_size_name(symtab,
+                                                                       arg) +
+                                                 "(:,:) => null()"]))
+            else:
+                parent.add(UseGen(parent, name="stencil_dofmap_mod", only=True,
+                                  funcnames=["stencil_dofmap_type"]))
+                parent.add(TypeDeclGen(parent, pointer=True,
+                                       datatype="stencil_dofmap_type",
+                                       entity_decls=[map_name+" => null()"]))
+                parent.add(DeclGen(parent, datatype="integer",
+                                   kind=api_config.default_kind["integer"],
+                                   pointer=True,
+                                   entity_decls=[self.dofmap_name(symtab, arg) +
+                                                 "(:,:,:) => null()"]))
+                parent.add(DeclGen(parent, datatype="integer",
+                                   kind=api_config.default_kind["integer"],
+                                   pointer=True,
+                                   entity_decls=[self.dofmap_size_name(symtab,
+                                                                       arg) +
+                                                 "(:) => null()"]))
+
             if stencil_type == "xory1d":
```

Comment:
Can you merge this if-block with the previous one?

Reply:
I've merged it to an extent, it still requires two if blocks but one is nested inside the other now and only required for stencils types other than cross2d.